### ISSUE TITLE
Support reading text by words and characters inside text fields on Android

### ISSUE THREAD

**Issue by gmoraja** [flutter/flutter#69794]
@gspencergoog

This happens inside text fields, with TalkBack on, and using a physical keyboard connected via Bluetooth.

TalkBack only switches between controls, links, headings, and paragraphs but doesn't allow users to browse by characters and words.

The problem has already been addressed in issue https://github.com/flutter/flutter/issues/13406 which is now closed, but we're still experiencing it.

**Comment by chunhtai** [comment]
I cannot reproduce it in my OnePlus, Do i need to have a physical keyboard connected?
```dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

void main() {
  runApp(Test());
}
int itemCount = 10;
class Test extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: '',
      routes: <String, WidgetBuilder>{
        '/': (_) => Scaffold(
          appBar: AppBar(title: const Text('Empty List Test')),
          body: itemCount > 0
            ? ListView.builder(
                itemCount: itemCount,
                itemBuilder: (BuildContext context, int index) {
                  return ListTile(
                    title: TextField(),
                  );
                },
              )
            : Center(child: const Text('No items')),
        ),
      },
    );
  }
}
```

**Comment by gmoraja** [comment]
@chunhtai yes, the problem only happens when browsing the content with a physical keyboard.

**Comment by chunhtai** [comment]
Hi @gmoraja  I still cannot reproduce this.


This is my repro step
1. make sure physical keyboard is connected through bluetooth
2. turn on talkback
3. run the app
```dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

void main() {
  runApp(Test());
}
int itemCount = 10;
class Test extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: '',
      routes: <String, WidgetBuilder>{
        '/': (_) => Scaffold(
          appBar: AppBar(title: const Text('Empty List Test')),
          body: itemCount > 0
            ? ListView.builder(
                itemCount: itemCount,
                itemBuilder: (BuildContext context, int index) {
                  return ListTile(
                    title: TextField(controller: TextEditingController(text: 'how are you'),),
                  );
                },
              )
            : Center(child: const Text('No items')),
        ),
      },
    );
  }
}
```
4. focus one text field and swipe up
5. I can reach all settings: control -> links-> heading -> paragraph -> word -> character, and they work correctly

**Comment by lorenzOliveto** [comment]
Hi @chunhtai , it's happening if you add a decoration with an hintText, and then navigate to the textField with the keyboard, with the Tab key.
We also need to wrap the TextField in a Semantics widget to give it a label, and use a FocusNode to intercept the arrow presses on the keyboard.

Here's an example Widget, and attached a video of the app running.

``` dart
import 'package:flutter/material.dart';

class KeyboardTbTextField extends StatefulWidget {
  @override
  _KeyboardTbTextFieldState createState() => _KeyboardTbTextFieldState();
}

class _KeyboardTbTextFieldState extends State<KeyboardTbTextField> {
  TextEditingController _textEditingController = TextEditingController();
  FocusNode _focusNode;

  @override
  void initState() {
    _focusNode = FocusNode(onKey: (node, key) {
      return true;
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Keyboard + TB + TextField')),
      body: Center(
        child: Semantics(
          label: 'Invite code',
          child: TextField(
            focusNode: _focusNode,
            decoration: InputDecoration(hintText: 'Enter text here'),
            controller: _textEditingController,
          ),
        ),
      ),
    );
  }
}
```

<details>

<summary>Flutter doctor output</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 1.24.0-8.0.pre.304, on macOS 11.0.1 20B29
    darwin-x64, locale it-IT)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
[✓] Android Studio (version 4.1)
[✓] Connected device (1 available)
    ! Error: iPhone di Alessandro is not connected. Xcode will continue when
      iPhone di Alessandro is connected. (code -13)

• No issues found!
```

</details>

[Video.zip](https://github.com/flutter/flutter/files/5565413/Video.zip)

**Comment by chunhtai** [comment]
I did a little research and this seems a WAI according to the android talkback service https://github.com/google/talkback/blob/9b1d5132b074174d02886faccf731e814d69363c/talkback/src/main/java/GranularityTraversal.java#L115

The granularity is disabled when the edit text has keyboard focus. I am not sure what is the reasoning behind their decision, but this is outside of the flutter's control.

It does not need a keyboard to reproduce this case. all you need to do is:

1. swipe to a textfield
2. double tap to keyboard focus
3. swipe left to focus something else
4. swipe right to refocus textfield
5. granularity is disabled.

The same thing happens for native android application. I am not sure why it requires step 3,4 for this to happen, it seems to be some bug inside talkback.

I will close this as WAI. if you think this is business critical, please reopen and I will see if I can reach out to android team for reason behind this decision.
